### PR TITLE
test: reproducers for store mutation with schema change and host down

### PR DESCRIPTION
**Pull request by alecco**
Reproducers for https://github.com/scylladb/scylladb/issues/10770.

(Already fixed in https://github.com/scylladb/scylladb/commit/15ebd5907177ba8709bc40ad2fa9ae773a9cf674)

Includes necessary improvements and fixes to `pylib`.

### CODE REVIEW

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/df25f23d889580b7153d84de74d1cefec286f5df:

- Removed change to default number primary keys as it breaks other tests
- commits RF reordered to be together

File: test/topology/test_mutation_schema_change.py
Comment (gleb-cloudius) [inline test/topology/test_mutation_schema_change.py:30]: Through C. The error only in the comment.

File: test/topology/test_mutation_schema_change.py
Comment (gleb-cloudius) [inline test/topology/test_mutation_schema_change.py:8]: The better description of the problem would be:
Reproducer for a failure during lwt operation due to missing of a column mapping in schema history table.

And please include the ref to the fix here 15ebd5907177ba8709bc40ad2fa9ae773a9cf674 which also have a longer description of the problem.

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/307fcc0f761c7321122b7587644274084a76fe3b:

- Fixed comments (Gleb)

File: test/topology/test_mutation_schema_change.py
Comment (gleb-cloudius) [inline test/topology/test_mutation_schema_change.py:62]: So did we decide in the end to go with sleeps?

**Review by gleb-cloudius — APPROVED** [review]

File: test/topology/test_mutation_schema_change.py
Comment (kostja) [reply test/topology/test_mutation_schema_change.py:62]: please no

**Comment by kbr-scylla** [comment]
If I understand correctly to eliminate the sleeps in this test we need to fix the alive nodes REST API?

How about for now we do what I suggested on Slack earlier:
>  you worked on having a way to reduce the group 0 snapshot threshold through error injection or some other means - I think we should have a simple test that does just this + a bunch of schema changes + adds a node to force a snapshot transfer, and simply expect that nothing breaks during the snapshot transfer. Maybe you could prepare such a test and send it as PR? We don't need the gossiper live nodes thing for that

So I suggest we start with a PR with a **simpler test** that only does the error injection for reducing snapshot threshold, performs a bunch of schema changes, and then bootstraps a node to force a snapshot transfer. This simpler test is not for reproducing https://github.com/scylladb/scylladb/issues/10770, it's only for checking that snapshot transfer works.

Separately we can work on fixing the alive nodes REST API.

Once both things are merged - the reducing snapshot threshold test and the REST API fix, we can go back to this PR and finish it without using sleeps.

**Comment by gleb-cloudius** [comment]
> If I understand correctly to eliminate the sleeps in this test we need to fix the alive nodes REST API?
> 
Yes.

> How about for now we do what I suggested on Slack earlier:
> 
> > you worked on having a way to reduce the group 0 snapshot threshold through error injection or some other means - I think we should have a simple test that does just this + a bunch of schema changes + adds a node to force a snapshot transfer, and simply expect that nothing breaks during the snapshot transfer. Maybe you could prepare such a test and send it as PR? We don't need the gossiper live nodes thing for that
> 
> So I suggest we start with a PR with a **simpler test** that only does the error injection for reducing snapshot threshold, performs a bunch of schema changes, and then bootstraps a node to force a snapshot transfer. This simpler test is not for reproducing #10770, it's only for checking that snapshot transfer works.

The test has nothing to do with raft. Why are we discussing snapshot transfer in its context?

**Comment by kbr-scylla** [comment]
> The test has nothing to do with raft. Why are we discussing snapshot transfer in its context?

The PR is introducing snapshot threshold error injection and enabling it in the test.

**Comment by gleb-cloudius** [comment]
> > The test has nothing to do with raft. Why are we discussing snapshot transfer in its context?
> 
> The PR is introducing snapshot threshold error injection and enabling it in the test.

This is needed for the test to work with raft enabled. But the test may as well ask to disable raft if it is possible instead. The test is not about snapshot transfer.

**Comment by kbr-scylla** [comment]
I want to have at least one test which forces snapshot transfer. We have 0 such tests right now. We should not be blocked from introducing a test that forces snapshot transfer by the alive endpoints issue because they are unrelated. Therefore I requested a separate PR with a test that specifically targets snapshot transfer which is not blocked by the alive endpoints issue (as this PR is) but which uses code from this PR (the error injection stuff). I requested it once on Slack and now repeating my request in this PR because this is the first PR that introduces the snapshot threshold error injection.

**Comment by gleb-cloudius** [comment]
> I want to have at least one test which forces snapshot transfer. We have 0 such tests right now. We should not be blocked from introducing a test that forces snapshot transfer by the alive endpoints issue because they are unrelated. Therefore I requested a separate PR with a test that specifically targets snapshot transfer which is not blocked by the alive endpoints issue (as this PR is) but which uses code from this PR (the error injection stuff). I requested it once on Slack and now repeating my request in this PR because this is the first PR that introduces the snapshot threshold error injection.

Makes sense.

**Comment by alecco** [comment]
> I want to have at least one test which forces snapshot transfer.

https://github.com/scylladb/scylladb/pull/12726

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/7ac9dd46d40cfe8464d334256c682526997d6049:

- rebased on top of https://github.com/scylladb/scylladb/pull/12540 (alive endpoints) (and that one is based on latest master)
- added helper for gossiper endpoints and removed sleeps
- snapshot error injection commits already merged in https://github.com/scylladb/scylladb/pull/12726

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/fc06b2acbdd446b5993bf9cccd83f8b495cd2e8c:

- rebased on latest https://github.com/scylladb/scylladb/pull/12540

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/856349f5bd81d52424a3ba6648fb27ccc35fe960:

- Parametrize replication factor for topology tests
- Keep using RF 1 for raft upgrade tests

@kbr-scylla

File: test/topology/conftest.py
Comment (kbr-scylla) [inline test/topology/conftest.py:221]: The default value for `replication_factor` in `RandomTables` is `1`, but there is only one use site of `RandomTables` and it's always passing `replication_factor` and the default is `3`.

Let's remove the default value inside `RandomTables` to make it less confusing.

File: test/topology/conftest.py
Comment (kbr-scylla) [inline test/topology/conftest.py:216]: Commit message: s/increase/decrease

File: test/topology_raft_disabled/test_mutation_schema_change.py
Comment (kbr-scylla) [inline test/topology_raft_disabled/test_mutation_schema_change.py:10]: I think running these tests in two suites is unnecessary. The downside is increasing the time to run all tests, in particular CI runs.

Also the tests you made use the `raft_server_snapshot_reduce_threshold` injection, so let's run them only in topology tests only.

**Review by kbr-scylla — CHANGES_REQUESTED** [review]
How often does each of these two tests reproduce the problem? Like how many times in 100 runs? Let's keep the one which reproduces more often and remove the other one.

File: test/topology/test_mutation_schema_change.py
Comment (gleb-cloudius) [reply test/topology/test_mutation_schema_change.py:90]: They are similar but trigger two different code paths.

**Comment by gleb-cloudius** [comment]
> How often does each of these two tests reproduce the problem? Like how many times in 100 runs? Let's keep the one which reproduces more often and remove the other one.

They trigger the bugs 100% on each run. They trigger two different cases though.

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/f464dab80c851a52d1149f4241b667b5afe0431f:

- rebased on https://github.com/scylladb/scylladb/pull/13240

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/f464dab80c851a52d1149f4241b667b5afe0431f:

- rebased on https://github.com/scylladb/scylladb/pull/13240

File: gms/gossiper.cc
Comment (alecco) [reply gms/gossiper.cc:761]: This pr was previously based on [another pr](https://github.com/scylladb/scylladb/pull/12540) that got merged on Friday.

But there was [an issue](https://github.com/scylladb/scylladb/issues/13235) over the weekend. Now there's [another gossiper pr](https://github.com/scylladb/scylladb/pull/13240) with a fix.

And this pr was rebased on top of that pending fix pr. This was mentioned in a comment earlier: https://github.com/scylladb/scylladb/pull/12699#issuecomment-1478254882

Comment (nyh) [reply gms/gossiper.cc:761]: Ok, I see. So we need to wait until this patch goes in and then you rebase this PR.

**Comment by alecco** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/

Restarted tests.

**Comment by alecco** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/
> 
> Restarted tests. (Edit: it was an unrelated failure)

Keeps failing on unrelated 
```
[backup_restore_tests.TestBackupRestore.test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/205/testReport/junit/backup_restore_tests/TestBackupRestore/Sanity_Tests___test_restore_snapshot_using_old_token_ownership/)
```

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/e322fbfb37df05f5da754cfeff5cbc3c682b305c:

- Reworded change of default replication factor (Kamil)
- Removed default replication factor parameter in `RandomTables` (Kamil)

**Comment by alecco** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/238/

Failed on unrelated debug upgrade test. Rebuilding.

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/b1484975971676f6ca9544856a43e6aa6fc6ae2f:

- Remove tests with raft disabled (Kamil)

Changes in https://github.com/scylladb/scylladb/commit/040da4e729d06ee8e85d6b37436f4f01792e6561:

- rebase to latest master

File: test/topology/test_mutation_schema_change.py
Comment (kbr-scylla) [inline test/topology/test_mutation_schema_change.py:74]: `[server_c.ip_addr]` is only the list of initial contact points for the driver: this is what `cluster_con` (in topology/conftest.py) does:
```python
    return Cluster(execution_profiles={EXEC_PROFILE_DEFAULT: profile},
                   contact_points=hosts,
```
`hosts` is the parameter.

The driver will automatically connect to all alive nodes after querying `system.peers` from the contact points.

So there's no guarantee that the lwt write will go through C as you said in the comment. It could go through A which is also alive now.

You'd have to specify a special `load_balancing_policy` in the `ExecutionProfile` passed to `Cluster` to make sure that the write goes through C.

File: test/topology/test_mutation_schema_change.py
Comment (kbr-scylla) [inline test/topology/test_mutation_schema_change.py:138]: Shouldn't we wait for `A` to see another server here, because we'll want to write through `A`?
(not sure I understand why this wait is needed)

File: test/topology/test_mutation_schema_change.py
Comment (kbr-scylla) [inline test/topology/test_mutation_schema_change.py:145]: See previous comment about contact points.

File: test/topology/test_mutation_schema_change.py
Comment (tgrabiec) [reply test/topology/test_mutation_schema_change.py:74]: I have a patch which allows you to connect to particular coordinators:

https://github.com/scylladb/scylladb/pull/13387/commits/4bea4ec1b71abc3ee9102f104c9d6d94351b42f3

Feel free to commit if good.

File: test/topology/test_mutation_schema_change.py
Comment (tgrabiec) [inline test/topology/test_mutation_schema_change.py:71]: This wait ensures that CQL requests don't fail, right? It's so generally useful, it should be an option to server_start().

File: test/topology/test_mutation_schema_change.py
Comment (kostja) [reply test/topology/test_mutation_schema_change.py:74]: @kbr-scylla @tgrabiec but what is the purpose here of going to a specific node? This is LWT, and if schema is broken on one node, we're going to have failures anyhow.  server_b is down, so LWT has to go to 2 nodes anyway. I don't see any reason to go to server_c specifically with this query.

File: test/topology/test_mutation_schema_change.py
Comment (kostja) [reply test/topology/test_mutation_schema_change.py:138]: It doesn't make any difference.

**Review by kostja — APPROVED** [review]

File: test/topology/test_mutation_schema_change.py
Comment (gleb-cloudius) [reply test/topology/test_mutation_schema_change.py:74]: We test very specific code path which require going to a specific node. We are not just running an LWT query to see it if succeeds.

Comment (kostja) [reply test/topology/test_mutation_schema_change.py:74]: I know but it doesn't require that it always goes to the same node. If it goes to a wrong node once in a while the test will succeed. Of course making the test more predictable is valuable, but even without it is good enough. That was my point. 

I tried at some point creating a connection pool that would set up a custom policy, this required very deep refactoring of the current manager code, since connections are opened on demand. 

Anyway @alecco please join the conversation and either address the comment or provide an update to the patch, I don't need to defend the patch for you I think.

Comment (gleb-cloudius) [reply test/topology/test_mutation_schema_change.py:74]: > I know but it doesn't require that it always goes to the same node. If it goes to a wrong node once in a while the test will succeed. 

Wont this just produce a fluky test?

Comment (alecco) [reply test/topology/test_mutation_schema_change.py]: Query goes to the proper node now.

File: test/topology/test_mutation_schema_change.py
Comment (alecco) [reply test/topology/test_mutation_schema_change.py]: Fixed in latest

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/eb4ffbc968cfe3537895ba83ddf789fae3f1d363:

- Use Tomek's [16075c5ceb](https://github.com/scylladb/scylladb/commit/4bea4ec1b71abc3ee9102f104c9d6d94351b42f3) to force driver to send query to the proper node (Tomek)
- Send query to prope node (Kamil)
- Wait for connections in proper node (Kamil)
- refactor `server_sees_other_server()` (to be moved to pylib later)

Changes in https://github.com/scylladb/scylladb/commit/fc43f10f7e31a6ed1edf1e862ca368c3d99ae0b6:

- rebase to latest master

File: test/pylib/random_tables.py
Comment (kbr-scylla) [inline test/pylib/random_tables.py]: The commit message says:

> When adding a new column it's always a value column, not clustering. Fix
the name and sequence.

no such thing is happening in this commit though.

File: test/topology/conftest.py
Comment (kbr-scylla) [inline test/topology/conftest.py]: `test_custom` needs to be fixed in a similar way in this `test/pylib: configurable replication factor` commit.

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/4639d5c97a8808f91b040803c051329229bb86b5:

- dropped commit "fix column name" as another version is already in master
- rebase on latest master

File: test/pylib/manager_client.py
Comment (kbr-scylla) [inline test/pylib/manager_client.py:51]: Please include a type hint: `server: Optional[ServerInfo] = None`

File: test/topology/test_mutation_schema_change.py
Comment (kbr-scylla) [inline test/topology/test_mutation_schema_change.py:75]: I don't understand why here you wait only for C to see another server, but in the other test you wait for both servers to see each other.

File: test/pylib/random_tables.py
Comment (alecco) [reply test/pylib/random_tables.py]: Yes, another commit was in master. I removed the commit in this series and rebased. Please latest.

File: test/pylib/manager_client.py
Comment (alecco) [reply test/pylib/manager_client.py:51]: This is Tomek's commit from the tablets series. I wanted to keep it as is so there's no conflict later. 

I'll fix it in [the subsequent PR](https://github.com/scylladb/scylladb/pull/13438).

File: test/topology/test_mutation_schema_change.py
Comment (alecco) [reply test/topology/test_mutation_schema_change.py:75]: The query here goes to C so it needs to know at least of 1 more to have QUORUM.

In the other location it goes through A so it needs to wait for it to know of 1 more for QUORUM. Perhaps the wait for C in the second case is not needed.

**Review by kbr-scylla — APPROVED** [review]

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/1c2ea5d6e090ed38b9daddc07f0b234929ddc03a:

- Re-added wait for C to see A in `test_mutation_schema_change_restart` (due to CI failures)

File: test/pylib/manager_client.py
Comment (alecco) [reply test/pylib/manager_client.py:51]: done

**Comment by alecco** [comment]
Changes in latest:
- Close driver before reconnecting

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/79ccbfc337dd44d06fa14148ec5030df505cbae8:

- use `wait_for_cql_and_get_hosts()` to avoid driver issue (Kamil)

Changes in https://github.com/scylladb/scylladb/commit/9597822214f891a27b9898f42602c62a677b0958:

- rebase on latest master

**Comment by kbr-scylla** [comment]
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/625/

reason for failure wasn't reported
@benipeled could you take a peek?

new build already running: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/632/

**Comment by benipeled** [comment]
> > CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/625/
> 
> reason for failure wasn't reported @benipeled could you take a peek?
> 

Failed on dtest https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/637/
The test-result visibility for dtest failures was improved by @tchaikov on https://github.com/scylladb/scylla-pkg/pull/3369 so it should be easier to figure out what's failed (on future runs)

**Pull request merged**
